src/dbally/llm_client/base.py

```diff
@@ -2,69 +2,79 @@
 
 from dbally.audit.event_tracker import EventTracker
 from dbally.data_models.audit import LLMEvent
-from dbally.data_models.llm_options import LLMOptions
 from dbally.prompts import ChatFormat, PromptBuilder, PromptTemplate
 
+LLMClientOptions = TypeVar("LLMClientOptions")
 
-class LLMClient(abc.ABC):
+
+@dataclass
+class LLMOptions(ABC):
+    """
+    Abstract dataclass that represents all available LLM call options.
+    """
+
+    dict = asdict
+
+
+class LLMClient(Generic[LLMClientOptions], ABC):
     """
     Abstract client for interaction with LLM.
 
-    It accepts parameters including the template, format, event tracker,
-    and optional generation parameters like frequency_penalty, max_tokens, and temperature
-    (the full list of options is provided by the [`LLMOptions` class][dbally.data_models.llm_options.LLMOptions]).
     It constructs a prompt using the `PromptBuilder` instance and generates text using the `self.call` method.
     """
 
-    def __init__(self, model_name: str):
+    _options_cls: Type[LLMClientOptions]
+
+    def __init__(self, model_name: str, default_options: Optional[LLMClientOptions] = None) -> None:
         self.model_name = model_name
+        self.default_options = default_options or self._options_cls()
         self._prompt_builder = PromptBuilder(self.model_name)
 
+    def __init_subclass__(cls) -> None:
+        if not hasattr(cls, "_options_cls"):
+            raise TypeError(f"Class {cls.__name__} is missing the '_options_cls' attribute")
+
     async def text_generation(  # pylint: disable=R0913
         self,
         template: PromptTemplate,
         fmt: dict,
         *,
         event_tracker: Optional[EventTracker] = None,
-        frequency_penalty: Optional[float] = 0.0,
-        max_tokens: Optional[int] = 128,
-        n: Optional[int] = 1,
-        presence_penalty: Optional[float] = 0.0,
-        seed: Optional[int] = None,
-        stop: Optional[Union[str, List[str]]] = None,
-        temperature: Optional[float] = 1.0,
-        top_p: Optional[float] = 1.0,
+        options: Optional[LLMClientOptions] = None,
     ) -> str:
         """
         For a given a PromptType and format dict creates a prompt and
         returns the response from LLM.
 
+        Args:
+            template: Prompt template in system/user/assistant openAI format.
+            fmt: Dictionary with formatting.
+            event_tracker: Event store used to audit the generation process.
+            options: options to use for the LLM client.
+
         Returns:
             Text response from LLM.
         """
-
-        options = LLMOptions(
-            frequency_penalty=frequency_penalty,
-            max_tokens=max_tokens,
-            n=n,
-            presence_penalty=presence_penalty,
-            seed=seed,
-            stop=stop,
-            temperature=temperature,
-            top_p=top_p,
-        )
+        options = options if options else self.default_options
```

Comment:
I believe the idea was for the options to be _merged_, so one can overwrite individual options, while leaving the rest as-is.

Reply:
Yes, I know, this was the case with the openai and anthropic integration, where they explicitly provide a sentinel for the `NotGiven` option, which helps in the implementation of merging. The problem is with the future integration with vertexai, which does not have a `NotGiven` sentinel, all parameters are optional and there is no way to distinguish if the user wanted to change a parameter or not, we would have to define a custom `NotGiven` sentinel and write a custom merge function for it. Therefore, I decided to override the options, I think this solution is cleaner, the user can easily modify default `LLMOptions` before passing them to `collection.ask` getting a similar effect.

Reply:
Ok, that makes sense, although in my mind it might also make sense to have our own provided-independent `NotGiven` that would be replaced with a provider-specific value at the very end (without exposing the provider-specific values to our end users)

Reply:
Good idea, added in #29.